src/haz3lmenhir/AST.re

```diff
@@ -0,0 +1,160 @@
+open Sexplib.Std;
+
+[@deriving (show({with_path: false}), sexp)]
+type filter_action =
+  | Pause
+  | Debug
+  | Hide
+  | Eval;
+
+[@deriving (show({with_path: false}), sexp)]
+type op_bin_float =
+  | Plus
+  | Minus
+  | Times
+  | Power
+  | Divide
+  | LessThan
+  | LessThanOrEqual
+  | GreaterThan
+  | GreaterThanOrEqual
+  | Equals
+  | NotEquals;
+
+[@deriving (show({with_path: false}), sexp)]
+type op_bin_bool =
+  | And
+  | Or;
+
+[@deriving (show({with_path: false}), sexp)]
+type op_bin_int =
+  | Plus
+  | Minus
+  | Times
+  | Power
+  | Divide
+  | LessThan
+  | LessThanOrEqual
+  | GreaterThan
+  | GreaterThanOrEqual
+  | Equals
+  | NotEquals;
+
+[@deriving (show({with_path: false}), sexp)]
+type op_bin_string =
+  | Concat
+  | Equals;
+
+[@deriving (show({with_path: false}), sexp)]
+type binOp =
+  | IntOp(op_bin_int)
+  | FloatOp(op_bin_float)
+  | StringOp(op_bin_string)
+  | BoolOp(op_bin_bool);
+
+[@deriving (show({with_path: false}), sexp)]
+type op_un_meta =
+  | Unquote;
+
+[@deriving (show({with_path: false}), sexp)]
+type op_un_int =
+  | Minus;
+
+[@deriving (show({with_path: false}), sexp)]
+type op_un_bool =
+  | Not;
+
+[@deriving (show({with_path: false}), sexp)]
+type op_un =
+  | Meta(op_un_meta)
+  | Int(op_un_int)
+  | Bool(op_un_bool);
+
+[@deriving (show({with_path: false}), sexp)]
+type typ_provenance =
+  | Internal
+  | EmptyHole;
+
+[@deriving (show({with_path: false}), sexp)]
+type typ =
+  | IntType
+  | StringType
+  | FloatType
+  | BoolType
+  | UnitType
+  | SumTyp(typ, option(typ))
```

Comment:
@green726   Instead of this being recursive with SumTerm can we have `SumTyp(list(string, option(typ))` for the variants?

Reply:
@7h3kk1d Defining the `SumTyp` with a `list` created a `reduce/reduce` error that I just couldn't seem to fix. I spent a while trying to do stuff with different lists and precedences and digging into the automaton but the only way I was able to resolve it was with this recursive style.

Reply:
I'll take a look as well but that makes sense.